docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,338 @@
+# Higher-order components
+
+A higher-order component (HOC) is an advanced technique in React for reusing component logic.
+
+Concretely, **a higher-order component is a function that returns a component.** An HOC accepts a component as one of its arguments, and returns an enhanced version of that component:
+
+```js
+let EnhancedComponent = higherOrderComponent(ChildComponent)
+```
+
+Whereas a component transforms props into UI, a higher-order component transforms one component into another component.
+
+HOCs are common in third-party React libraries, such as Redux's `connect()` and Relay's `createContainer()`.
+
+In this document, we'll discuss why higher-order components are useful, and how to write your own.
+
+## Use HOCs for cross-cutting concerns
+
+Components are the primary unit of code reuse in React. However, you'll find that some patterns aren't easily abstracted into traditional components.
```

Comment:
This makes it sound like we claim render callbacks are "hard".
Maybe "sometimes it is harder to figure out how to abstract some pattern into a traditional component".